libs/compute_cuda/include/hpx/compute/cuda/target.hpp

```diff
@@ -32,6 +35,149 @@
+                HPX_ASSERT(rt);
+
+                // Register this thread with HPX, this should be done once for
+                // each external OS-thread intended to invoke HPX functionality.
+                // Calling this function more than once on the same thread will
+                // report an error.
+                hpx::error_code ec(hpx::lightweight);    // ignore errors
+                hpx::register_thread(rt_, "cuda", ec);
+            }
+            ~runtime_registration_wrapper()
+            {
+                // Unregister the thread from HPX, this should be done once in
+                // the end before the external thread exists.
+                hpx::unregister_thread(rt_);
+            }
+
+            hpx::runtime* rt_;
+        };
+
+        template <typename Allocator>
+        struct future_data;
+
+        template <typename Allocator>
+        struct release_on_exit
+        {
+            release_on_exit(future_data<Allocator>* data)
+              : data_(data)
+            {
+            }
+
+            ~release_on_exit()
+            {
+                // release the shared state
+                lcos::detail::intrusive_ptr_release(data_);
+            }
+
+            future_data<Allocator>* data_;
+        };
+
+        template <typename Allocator>
+        struct future_data
+          : lcos::detail::future_data_allocator<void, Allocator>
+        {
+            using init_no_addref =
+                typename lcos::detail::future_data_allocator<void,
+                    Allocator>::init_no_addref;
+            using other_allocator = typename std::allocator_traits<
+                Allocator>::template rebind_alloc<future_data>;
+
+            static void CUDART_CB stream_callback(
+                cudaStream_t stream, cudaError_t error, void* user_data)
+            {
+                future_data* this_ = static_cast<future_data*>(user_data);
+
+                runtime_registration_wrapper wrap(this_->rt_);
+                release_on_exit<Allocator> on_exit(this_);
+
+                if (error != cudaSuccess)
+                {
+                    this_->set_exception(HPX_GET_EXCEPTION(kernel_error,
+                        "cuda::detail::future_data::stream_callback()",
+                        std::string("cudaStreamAddCallback failed: ") +
+                            cudaGetErrorString(error)));
+                    return;
+                }
+
+                this_->set_data(hpx::util::unused);
+            }
+
+            future_data()
+              : rt_(hpx::get_runtime_ptr())
+            {
+            }
+
+            future_data(init_no_addref no_addref, other_allocator const& alloc,
+                cudaStream_t stream)
+              : lcos::detail::future_data_allocator<void, Allocator>(
+                    no_addref, alloc)
+              , rt_(hpx::get_runtime_ptr())
+            {
+                init(stream);
+            }
+
+            void init(cudaStream_t stream)
```

Comment:
This function could be HPX_EXPORT'ed and moved to the .cpp file

Reply:
This function can't be HPX_EXPORT'ed and moved to the implementation file after all. It's a non-static member function of a template.